### ISSUE TITLE
`vue/require-default-prop` should not trigger when type is `x | undefined` without default value

### ISSUE THREAD

**Issue by Aspyryan** [vuejs/eslint-plugin-vue#2475]
**What rule do you want to change?**
`vue/require-default-prop`

**Does this change cause the rule to produce more or fewer warnings?**
Is will provide less warnings if the type is already defined as undefined.

**How will the change be implemented? (New option, new default behavior, etc.)?**
Could be an option to not throw the error / warning when props is not required and has a undefined type set

**Please provide some example code that this change will affect:**
```ts
const props = withDefaults(
  defineProps<{
    // Is not throwing warning because default is set
    modelValue?: string;
    // Should throw warning because no default is set
    label?: string;
    // Should not throw warning because undefined is allowed
    propertyName?: string | undefined;
  }>(),
  {
    modelValue: 'test',
  },
);
```

**What does the rule currently do for this code?**
Throws a warning at the propertyName prop

**What will the rule do after it's changed?**
Will not complain about the propertyName prop

**Additional context**
If needed, I can look into it myself but I will need to be pointed to the correct files :)

**Comment by FloEdelmann** [comment]
Related to

* https://github.com/vuejs/eslint-plugin-vue/issues/2051
* https://github.com/vuejs/eslint-plugin-vue/issues/2376

**Comment by FloEdelmann** [comment]
I'm not sure if I agree; every optional property (with a `?`) automatically has `| undefined` added to its type, so it would render the rule useless if every optional prop would not be required to have a default value then.

**Comment by Aspyryan** [comment]
Hmm yea I get that, is it typescript doing the automatic undefined? can we somehow disable that and then apply the rule?

**Comment by FloEdelmann** [comment]
Yes it is. I don't think we can turn that off.

**Comment by higherorderfunctor** [comment]
https://www.typescriptlang.org/tsconfig/#exactOptionalPropertyTypes

I'm using it in my projects to force an explicit `undefined` on optional properties where expected.

Not sure if it helps with this exact situation, but there is support in TS for it.

One of our dependencies requires it: https://github.com/Effect-TS/effect/tree/main/packages/schema#requirements.

**Comment by mrleblanc101** [comment]
Just updated from 10.0.0 where I had no eslint error to 10.1.0, and now I have `vue/require-default-prop` popping everywhere.

Before:
<img width="816" alt="Image" src="https://github.com/user-attachments/assets/caae0c26-67bc-46e0-9c06-d22b9f181900" />

After:
<img width="823" alt="Image" src="https://github.com/user-attachments/assets/5164b8f4-1891-4f86-8348-a7cfea71d24f" />

**Comment by Perdolique** [comment]
New typescript-eslint rule [`no-useless-default-assignment`](https://typescript-eslint.io/rules/no-useless-default-assignment/) is conflicting with `vue/require-default-prop` and optional props like

```ts
interface Props {
  readonly text?: string;
}

const {
  text = undefined // <- "no-useless-default-assignment" error with this assignment, or "vue/require-default-prop" without
} = defineProps<Props>();
```
